### PR TITLE
feat: Add fine tuning spec in CRD

### PR DESCRIPTION
**Pull request by Fei-Guo**
This PR contains the change of having the workspace CRD to support model fine tuning. The design follows the pattern of interference spec such that the complex training arguments are not exposed via the CRD API. Instead, a default configmap will be provided and users can override it with customization.

### CODE REVIEW

File: api/v1alpha1/workspace_types.go [hunk 1]
```diff
@@ -73,7 +73,7 @@ type PresetSpec struct {
 }
 
 type InferenceSpec struct {
-	// Preset describles the model that will be deployed with preset configurations.
+	// Preset describles the base model that will be deployed with preset configurations.
```
Comment (ishaansehgal99) [inline api/v1alpha1/workspace_types.go:76]: describes

File: api/v1alpha1/workspace_types.go [hunk 2]
```diff
@@ -83,6 +83,78 @@ type InferenceSpec struct {
...
+	Strength *string `json:"strength,omitempty"`
+}
+
+type DataSource struct {
+	// The name of the dataset. The same name will be used as a container name.
+	// It must be a valid DNS subdomain value,
+	Name string `json:"name,omitempty"`
+	// URLs specifies the links to the public data sources. E.g., files in a public github repository.
+	// +optional
+	URLs []string `json:"uRLs,omitempty"`
```
Comment (ishaansehgal99) [inline api/v1alpha1/workspace_types.go:108]: uRLs -> should either be URLs or urls

**Review by ishaansehgal99 — APPROVED** [review]

File: api/v1alpha1/workspace_types.go [hunk 3]
```diff
@@ -83,6 +83,78 @@ type InferenceSpec struct {
...
+	// The directory in the host that contains the output data.
+	// +optional
+	HostPath string `json:"hostPath,omitempty"`
+	// Name of the image where the output data is pushed to.
+	// +optional
+	Image string `json:"image,omitempty"`
+	// ImagePushSecret is the name of the secret in the same namespace that contains the authentication
+	// information that is needed for running `docker push`.
+	// +optional
+	ImagePushSecret string `json:"imagePushSecret,omitempty"`
```
Comment (ishaansehgal99) [inline api/v1alpha1/workspace_types.go:131]: ImagePullSecrets is an array should this also be an array

File: api/v1alpha1/workspace_types.go (hunk 1 above)
Comment (Fei-Guo) [reply api/v1alpha1/workspace_types.go:76]: done

File: api/v1alpha1/workspace_types.go (hunk 2 above)
Comment (Fei-Guo) [reply api/v1alpha1/workspace_types.go:108]: done

File: api/v1alpha1/workspace_types.go [hunk 4]
```diff
@@ -83,6 +83,78 @@ type InferenceSpec struct {
...
+	// Config specifies the name of the configmap in the same namespace that contains the arguments used by the tuning method.
+	// If not specified, a default configmap is used based on the specified method.
+	// +optional
+	Config string `json:"config,omitempty"`
+	// Input describes the input used by the tuning method.
+	// +optional
+	Input *DataSource `json:"input,omitempty"`
+	// Output specified where to store the tuning output.
+	// +optional
+	Output *DataDestination `json:"output,omitempty"`
```
[the remark this answers is not in the document]
Comment (Fei-Guo) [reply api/v1alpha1/workspace_types.go:157]: Make them required.

**Pull request merged**